indra/llconvexdecomposition/llconvexdecompositionvhacd.cpp

```diff
@@ -0,0 +1,492 @@
+    {
+        return LLCD_NULL_PTR;
+    }
+
+    return mBoundDecomp->mSourceMesh.from(data, vertex_based);
+}
+
+LLCDResult LLConvexDecompositionVHACD::registerCallback(int stage, llcdCallbackFunc callback )
+{
+    if (stage == 0)
+    {
+        mVHACDCallback.setCallbackFunc(callback);
+        return LLCD_OK;
+    }
+    else
+    {
+        return LLCD_INVALID_STAGE;
+    }
+}
+
+LLCDResult LLConvexDecompositionVHACD::executeStage(int stage)
+{
+    if (!mBoundDecomp)
+    {
+        return LLCD_NULL_PTR;
+    }
+
+    if (stage != 0)
+    {
+        return LLCD_INVALID_STAGE;
+    }
+
+    mBoundDecomp->mDecomposedHulls.clear();
+
+    const auto& decomp_mesh = mBoundDecomp->mSourceMesh;
+    if (!mVHACD->Compute((const double* const)decomp_mesh.mVertices.data(), static_cast<uint32_t>(decomp_mesh.mVertices.size()), (const uint32_t* const)decomp_mesh.mIndices.data(), static_cast<uint32_t>(decomp_mesh.mIndices.size()), mVHACDParameters))
+    {
+        return LLCD_INVALID_HULL_DATA;
+    }
+
+    uint32_t num_nulls = mVHACD->GetNConvexHulls();
+    if (num_nulls == 0)
+    {
+        return LLCD_INVALID_HULL_DATA;
+    }
+
+    for (uint32_t i = 0; num_nulls > i; ++i)
+    {
+        VHACD::IVHACD::ConvexHull ch;
+        if (!mVHACD->GetConvexHull(i, ch))
+            continue;
+
+        LLConvexMesh out_mesh;
+        out_mesh.setVertices(ch.m_points);
+        out_mesh.setIndices(ch.m_triangles);
+
+        mBoundDecomp->mDecomposedHulls.push_back(std::move(out_mesh));
+    }
+
+    mVHACD->Clean();
+
+    return LLCD_OK;
+}
+
+LLCDResult LLConvexDecompositionVHACD::buildSingleHull()
+{
+    LL_INFOS() << "Building single hull mesh" << LL_ENDL;
+    if (!mBoundDecomp || mBoundDecomp->mSourceMesh.mVertices.empty())
+    {
+        return LLCD_NULL_PTR;
+    }
+
+    mBoundDecomp->mSingleHullMesh.clear();
+
+    VHACD::QuickHull quickhull;
+    uint32_t num_tris = quickhull.ComputeConvexHull(mBoundDecomp->mSourceMesh.mVertices, MAX_VERTICES_PER_HULL);
+    if (num_tris > 0)
+    {
+        mBoundDecomp->mSingleHullMesh.setVertices(quickhull.GetVertices());
+        mBoundDecomp->mSingleHullMesh.setIndices(quickhull.GetIndices());
+
+        return LLCD_OK;
+    }
+
+    return LLCD_INVALID_MESH_DATA;
+}
+
+int LLConvexDecompositionVHACD::getNumHullsFromStage(int stage)
+{
+    if (!mBoundDecomp || stage != 0)
+    {
+        return 0;
+    }
+
+    return narrow(mBoundDecomp->mDecomposedHulls.size());
+}
+
+LLCDResult LLConvexDecompositionVHACD::getSingleHull( LLCDHull* hullOut )
+{
+    memset( hullOut, 0, sizeof(LLCDHull) );
+
+    if (!mBoundDecomp)
+    {
+        return LLCD_NULL_PTR;
+    }
+
+    if (mBoundDecomp->mSingleHullMesh.vertices.empty())
+    {
+        return LLCD_INVALID_HULL_DATA;
+    }
+
+    mBoundDecomp->mSingleHullMesh.to(hullOut);
+    return LLCD_OK;
+}
+
+LLCDResult LLConvexDecompositionVHACD::getHullFromStage( int stage, int hull, LLCDHull* hullOut )
+{
+    memset( hullOut, 0, sizeof(LLCDHull) );
```

Comment:
Functionally the same, but these memsets would be cleaner and more readable like
```suggestion
    *hullOut = LLCDHull{};
```